prow/googlecloudbuild/client/client.go

```diff
@@ -0,0 +1,182 @@
+	parts := strings.SplitN(tag, ProwLabelSeparater, 2)
+	if len(parts) < 2 {
+		return tag, ""
+	}
+	return parts[0], parts[1]
+}
+
+// GetProwLabels gets labels from cloud build struct, simulating k8s pods labels format
+// of map[string]string
+func GetProwLabels(bld *cloudbuildpb.Build) map[string]string {
+	res := make(map[string]string)
+	for _, tag := range bld.Tags {
+		key, val := LabelFromTag(tag)
+		if val != "" {
+			res[key] = val
+		}
+	}
+	return res
+}
+
+// Client wraps native cloudbuild client
+type Client struct {
+	interactor *cloudbuild.Client
+}
+
+// NewClient creates a new Client, with optinal credentialFile
+func NewClient(ctx context.Context, credentialFile string) (*Client, error) {
+	var opts []option.ClientOption
+	// Authenticating with key file if it's provided
+	if len(credentialFile) > 0 {
+		if _, err := os.Stat(credentialFile); err != nil {
+			return nil, err
+		}
+		opts = append(opts, option.WithCredentialsFile(credentialFile))
+	}
+
+	cbClient, err := cloudbuild.NewClient(ctx, opts...)
+	if err != nil {
+		return nil, err
+	}
+	return &Client{interactor: cbClient}, nil
+}
+
+// GetBuild gets build by it's id
+func (c *Client) GetBuild(ctx context.Context, project, id string) (*cloudbuildpb.Build, error) {
+	return c.interactor.GetBuild(ctx, &cloudbuildpb.GetBuildRequest{
+		ProjectId: project,
+		Id:        id,
+	})
+}
+
+// ListBuildsByTag lists builds by tags
+func (c *Client) ListBuildsByTag(ctx context.Context, project string, tags []string) ([]*cloudbuildpb.Build, error) {
+	const pageSize = 50
+	var res []*cloudbuildpb.Build
+	var tagsFilters []string
+	for _, tag := range tags {
+		tagsFilters = append(tagsFilters, "tags="+tag)
+	}
+
+	iter := c.interactor.ListBuilds(ctx, &cloudbuildpb.ListBuildsRequest{
+		ProjectId: project,
+		PageSize:  pageSize,
+		Filter:    strings.Join(tagsFilters, " AND "),
+	})
+	// ListBuilds already fetches all, just need to do pagination
+	pager := iterator.NewPager(iter, pageSize, "")
+	for {
+		var buildsInPage []*cloudbuildpb.Build
+		nextPageToken, err := pager.NextPage(&buildsInPage)
+		if err != nil {
+			return nil, err
+		}
+		if buildsInPage != nil {
+			res = append(res, buildsInPage...)
+		}
+		if nextPageToken == "" {
+			break
+		}
+	}
+	return res, nil
+}
+
+// CreateBuild creates build and wait for the operation to complete.
+func (c *Client) CreateBuild(ctx context.Context, project string, bld *cloudbuildpb.Build) (*cloudbuildpb.Build, error) {
+	op, err := c.interactor.CreateBuild(ctx, &cloudbuildpb.CreateBuildRequest{
+		ProjectId: project,
+		Build:     bld,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// CreateBuild returns CreateBuildOperation, wait until the operation results in build.
+	// CreateBuildOperation contains `Wait` method, which however polls every minute, which is
+	// too slow. So use `wait.Poll` instead.
+	var triggered *cloudbuildpb.Build
+	if err := wait.Poll(100*time.Millisecond, 30*time.Second, func() (bool, error) {
```

Comment:
I think 100*time.Millisecond should be a a named constant instead. But also, is this 0.10 second interval too short? I fear getting throttled by GCB if we're doing this across possibly hundreds of pending builds.

Also, what does the `30*time.Second` interval do? Can you name this constant (meaningfully) as well and use it here?

Reply:
done